crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,205 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::TranspilerError;
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use qiskit_circuit::dag_node::DAGNode;
+use qiskit_circuit::operations::{OperationRef, StandardInstruction};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::petgraph::prelude::NodeIndex;
+
+pub fn run_alap_schedule_analysis(
+    dag: &DAGCircuit,
+    clbit_write_latency: u64,
+    node_durations: HashMap<NodeIndex, f64>,
+) -> PyResult<HashMap<NodeIndex, f64>> {
+    if dag.qregs().len() != 1 || !dag.qregs_data().contains_key("q") {
+        return Err(TranspilerError::new_err(
+            "ALAP schedule runs on physical circuits only",
+        ));
+    }
+
+    let mut node_start_time: HashMap<NodeIndex, f64> = HashMap::new();
+    let mut idle_before: HashMap<Wire, f64> = HashMap::new();
+
+    for index in 0..dag.qubits().len() {
+        idle_before.insert(Wire::Qubit(Qubit::new(index)), 0.0);
+    }
+
+    for index in 0..dag.clbits().len() {
+        idle_before.insert(Wire::Clbit(Clbit::new(index)), 0.0);
+    }
+
+    // Since this is alap scheduling, node is scheduled in reversed topological ordering
+    // and nodes are packed from the very end of the circuit.
+    // The physical meaning of t0 and t1 is flipped here.
+
+    for node_index in dag
+        .topological_op_nodes()?
+        .collect::<Vec<_>>()
+        .into_iter()
+        .rev()
+    {
+        let op = match dag.dag().node_weight(node_index) {
+            Some(NodeType::Operation(op)) => op,
+            _ => panic!("topological_op_nodes() should only return instances of DagOpNode."),
+        };
```

Comment:
I think that now you should be able to access the `DAG` via indexing to get the node's weight (`NodeType`). And you also have the `unwrap_operation()` to get the PackedInstruction.


```suggestion
        let op = dag[node_index].unwrap_operation();
```